### PR TITLE
Enable dependency analyzer in fail mode in order to automatically detect dependency problems

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

Enable dependency analyzer in fail mode.

The changes boil down to:

1. some dependencies had the wrong scope (compile rather than test, mostly)
1. some dependencies were superfluous. 
1. some direct dependencies were missing
    * [unfortunately testing-impl](https://github.com/kroxylicious/kroxylicious-junit5-extension/issues/271) is a direct dependency in a few places (including the sample 😔).
    * the sample also has a test scope dependency on kroxylicious-runtime (used to construct the filter config).
1. hamcrest is a problem - we were ending up with version 1 and 2 on the classpath.
    _the neatest solution I found was the hamcrest-core exclusions you see in integration-tests and integration-test-support_.  Here's why it is hard..
    * test-containers still transitivity depends on hamcrest 1.3 (via junit 4) - https://github.com/testcontainers/testcontainers-java/issues/970
    * awaitility depends on hamcest 2
    * hamcrest 2 and hamcrest 1 use different maven artefact names.
1. introducing the `analyse-only` task upset the execution order which broke krpc generation (smelt of a maven bug... https://issues.apache.org/jira/browse/MNG-5799 maybe.. time sink... step away).  That's why you see the lifecycling binding change to keep the execution order: unpack then generate.
 
_Please describe your pull request_

### Additional Context

why: automate the detection of unused declared dependencies and used undeclared dependencies at build time

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-api/pom.xml [hunk 1]
```diff
@@ -39,11 +39,6 @@
             <groupId>com.github.spotbugs</groupId>
             <artifactId>spotbugs-annotations</artifactId>
         </dependency>
-
-        <dependency>
-            <groupId>io.micrometer</groupId>
-            <artifactId>micrometer-core</artifactId>
```
Comment (robobario) [inline kroxylicious-api/pom.xml:45]: might be worth a note in the changelog that the API had a dependency removed, similar with the kroxylicious-kms deps.

**Review by robobario — APPROVED** [review]
LGTM. the hamcrest situation :cry:

**Comment by tombentley** [comment]
Do hamcrest1 and 2 have any overlapping package names?

**Comment by robobario** [comment]
looks like the sonar build is failing, maybe one of the dependency detections fails with ITs switched off?

**Comment by k-wall** [comment]
> Do hamcrest1 and 2 have any overlapping package names?

Yes.   Using Hamcrest 2 jars instead of Hamcrest 1 works for our use-case.

**Comment by k-wall** [comment]
> looks like the sonar build is failing, maybe one of the dependency detections fails with ITs switched off?

```
[INFO] --- failsafe:3.2.5:integration-test (integration-test) @ kroxylicious-integration-tests ---
[INFO] Tests are skipped.
[INFO]
[INFO] --- dependency:3.6.1:analyze-only (analyze) @ kroxylicious-integration-tests ---
[ERROR] Unused declared dependencies found:
[ERROR]    io.kroxylicious:kroxylicious-api:jar:0.5.0-SNAPSHOT:test
[ERROR]    io.kroxylicious:kroxylicious-runtime:jar:0.5.0-SNAPSHOT:test
[ERROR]    io.kroxylicious:kroxylicious-integration-test-support:jar:0.5.0-SNAPSHOT:test
[ERROR]    io.kroxylicious:kroxylicious-kms-test-support:jar:0.5.0-SNAPSHOT:test
[ERROR]    io.kroxylicious:kroxylicious-encryption:jar:0.5.0-SNAPSHOT:test
[ERROR]    io.kroxylicious:kroxylicious-kms-provider-kroxylicious-inmemory:jar:0.5.0-SNAPSHOT:test
[ERROR]    io.kroxylicious:kroxylicious-multitenant:jar:0.5.0-SNAPSHOT:test
[ERROR]    io.kroxylicious:kroxylicious-record-validation:jar:0.5.0-SNAPSHOT:test
[ERROR]    io.kroxylicious:kroxylicious-simple-transform:jar:0.5.0-SNAPSHOT:test
[ERROR]    io.kroxylicious.testing:testing-api:jar:0.8.1:test
[ERROR]    io.kroxylicious.testing:testing-junit5-extension:jar:0.8.1:test
[ERROR]    io.kroxylicious.testing:testing-impl:jar:0.8.1:test
[ERROR]    com.github.spotbugs:spotbugs-annotations:jar:4.8.3:provided
[ERROR]    com.fasterxml.jackson.core:jackson-annotations:jar:2.16.1:test
[ERROR]    io.micrometer:micrometer-core:jar:1.12.1:test
[ERROR]    io.netty:netty-codec-http:jar:4.1.104.Final:test
[ERROR]    io.netty:netty-common:jar:4.1.104.Final:test
[ERROR]    org.apache.kafka:kafka-clients:jar:3.6.1:test
[ERROR]    org.apache.logging.log4j:log4j-api:jar:2.22.1:test
[ERROR]    org.apache.logging.log4j:log4j-core:jar:2.22.1:test
[ERROR]    org.awaitility:awaitility:jar:4.2.0:test
[ERROR]    org.junit.jupiter:junit-jupiter-api:jar:5.10.1:test
[ERROR]    org.junit.jupiter:junit-jupiter-params:jar:5.10.1:test
[ERROR]    org.assertj:assertj-core:jar:3.25.1:test
[ERROR]    org.hamcrest:hamcrest:jar:2.2:test
[ERROR]    org.slf4j:slf4j-api:jar:2.0.11:test
```

**Comment by SamBarker** [comment]
> edit: I think a jdk activation would be better solution https://maven.apache.org/guides/introduction/introduction-to-profiles.html#jdk than the property based activation we have today, but that can be separate commit.

I'm not sure JDK activation does exactly the same thing. It would stop us trying to build run integration tests if JDK21 is not in use, but that would mean we run the tests, for no value, when running the sonar scan. Having said that JDK activation probably makes more sense for the original requirement....

**Pull request merged**